File: AdventOfCode2024-CSharp/src/AdventOfCode2024.Puzzles/Report.cs
using System;
using System.Collections.Generic;

namespace AdventOfCode2024;

internal static class Report
{
    internal static Report<TLevels> Create<TLevels>(TLevels levels, int excludedIndex)
        where TLevels : IReadOnlyList<int> =>
        new(levels, excludedIndex);
}

internal readonly record struct Report<TLevels>(TLevels Levels, int ExcludedIndex) where TLevels : IReadOnlyList<int>
{
    private bool TryDecrement(int index, out int decrementedIndex)
    {
        int candidate = index - 1;
        decrementedIndex = candidate == ExcludedIndex ? candidate - 1 : candidate;
        return decrementedIndex >= 0;
    }

    private bool TryIncrement(int index, out int incrementedIndex)
    {
        int candidate = index + 1;
        incrementedIndex = candidate == ExcludedIndex ? candidate + 1 : candidate;
        return incrementedIndex < Levels.Count;
    }

    internal bool IsLevelSafeLeft(int index, out int sign)
    {
        if (!TryDecrement(index, out int previousIndex))
        {
            sign = default;
            return true;
        }

        int signedDifference = Levels[index] - Levels[previousIndex];
        sign = Math.Sign(signedDifference);
        return Math.Abs(signedDifference) is >= 1 and <= 3;
    }

    internal bool IsLevelSafeRight(int index, out int sign)
    {
        if (!TryIncrement(index, out int nextIndex))
        {
            sign = default;
            return true;
        }

        int signedDifference = Levels[nextIndex] - Levels[index];
        sign = Math.Sign(signedDifference);
        return Math.Abs(signedDifference) is >= 1 and <= 3;
    }
}

File: AdventOfCode2024-Kotlin/src/main/kotlin/Report.kt
package com.adventofcode2024

import kotlin.math.abs
import kotlin.math.sign

data class Report<TLevels>(val levels: TLevels, val excludedIndex: Int) where TLevels : List<Int> {

    companion object {
        fun <TLevels> create(levels: TLevels, excludedIndex: Int): Report<TLevels> where TLevels : List<Int> =
            Report(levels, excludedIndex)
    }

    private fun tryDecrement(index: Int): Int? {
        val candidate = index - 1
        return (if (candidate == excludedIndex) candidate - 1 else candidate).takeIf { it >= 0 }
    }

    private fun tryIncrement(index: Int): Int? {
        val candidate = index + 1
        return (if (candidate == excludedIndex) candidate + 1 else candidate).takeIf { it < levels.size }
    }

    fun tryGetLeftSignIfSafe(index: Int): Int? {
        val previousIndex = tryDecrement(index) ?: return 0
        val signedDifference = levels[index] - levels[previousIndex]
        return if (abs(signedDifference) in 1..3) signedDifference.sign else null
    }

    fun tryGetRightSignIfSafe(index: Int): Int? {
        val nextIndex = tryIncrement(index) ?: return 0
        val signedDifference = levels[nextIndex] - levels[index]
        return if (abs(signedDifference) in 1..3) signedDifference.sign else null
    }

}
